Lay groundwork for multiple lists in shopping-list

iantrich opened the pull request:
## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

iantrich committed:
Lay groundwork for multiple lists in shopping-list

iantrich committed:
lint

iantrich committed:
Remove logger

iantrich commented:
Need to add a test still

iantrich committed:
Added a test

iantrich committed:
lint

balloob commented on homeassistant/components/shopping_list.py:
```diff
@@ -34,36 +35,47 @@
 SERVICE_ITEM_SCHEMA = vol.Schema({
-    vol.Required(ATTR_NAME): vol.Any(None, cv.string)
+    vol.Required(ATTR_NAME): vol.Any(None, cv.string),
+    vol.Optional(ATTR_LIST_ID, '0'): cv.string
```
I don't think that we should have our IDs be just numbers or have `'0'` be the default one. I think that it really depends on what lists are available. Would just not give it a default value.

iantrich replied:
Default of '0' is so the existing panel will still work and voice intents and I know what it is.

balloob replied:
Is the idea that future lists will be 1, 2, 3 etc ? And if I remove 2, will there be a hole or will all subsequent IDs switch one back?

I think that I would prefer the following approach:

- Convert the current list to a system generated list named by default "Inbox" that the user cannot delete. The ID will be hardcoded to `system-inbox`. 

The ids of all other lists will be generated using `uuid.uuid4().hex`

iantrich replied:
No, my plan is actually exactly the approach you stated. I should have made that clear, besides the id being `system-inbox`, I like that change.

iantrich replied:
@balloob do you think I should have a way to allow current shopping-list users to move their current items or is clearing them okay if listed as a breaking change?

balloob commented on homeassistant/components/shopping_list.py:
```diff
@@ -34,36 +35,47 @@
 SCHEMA_WEBSOCKET_UPDATE_ITEM = \
     websocket_api.BASE_COMMAND_MESSAGE_SCHEMA.extend({
         vol.Required('type'): WS_TYPE_SHOPPING_LIST_UPDATE_ITEM,
+        vol.Required('list_id'): str,
```
Since I don't see an accompanying frontend PR, this would break the frontend if we merge this.

iantrich replied:
https://github.com/home-assistant/home-assistant-polymer/pull/2164

balloob commented on homeassistant/components/shopping_list.py:
```diff
@@ -175,23 +211,32 @@ def async_update(self, item_id, info):
         def load():
             """Load the items synchronously."""
-            return load_json(self.hass.config.path(PERSISTENCE), default=[])
+            return load_json(self.hass.config.path(PERSISTENCE), default=[{
```
If we're migrating to multiple lists, we should also migrate to the Storage helper. 

Here is an example of how to migrate existing files to the storage helper https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity_registry.py#L206-L210

iantrich replied:
Is this so it will be stored in `.storage`? Agreed. Thanks.

balloob replied:
Yep

iantrich replied:
I'll try to figure this out tonight.

iantrich committed:
Update to use system-id as default id

balloob commented on homeassistant/components/shopping_list.py:
```diff
@@ -36,7 +36,7 @@
 SERVICE_ITEM_SCHEMA = vol.Schema({
     vol.Required(ATTR_NAME): vol.Any(None, cv.string),
-    vol.Optional(ATTR_LIST_ID, '0'): cv.string
+    vol.Optional(ATTR_LIST_ID, 'system-inbox'): cv.string
```
Let's store this in a constant.

iantrich committed:
Extract const

balloob commented on homeassistant/components/shopping_list.py:
```diff
@@ -15,9 +15,11 @@
 ATTR_NAME = 'name'
 
 DOMAIN = 'shopping_list'
+DEFAULT_LIST = 'system-inbox'
```
Can we rename this `LIST_SYSTEM_INBOX` ? We should name things to what they are, not how they are used, because it can change how it's used in the future. It's like `report_final_final_final_v2.pdf` 😉

iantrich committed:
Update const name

iantrich committed:
Migration: Has some issues that I need to resolve

iantrich commented:
@balloob no need to look at this yet, there are still some pending issues. Thanks.

iantrich committed:
Fixed tests and migration function

iantrich commented:
I know you're a busy man, but just wanted to let you know that I believe this is in a good state now. Thanks.

iantrich committed:
Address some review comments

iantrich commented on homeassistant/components/shopping_list.py:
```diff
@@ -175,23 +215,34 @@ def async_update(self, item_id, info):
 
-        self.items = yield from self.hass.async_add_job(load)
+        lis['items'] = [itm for itm in lis['items'] if not itm['complete']]
+        self.hass.async_add_job(self.save)
```
I believe I addressed this with `asynce_delay_save`

iantrich commented on homeassistant/components/shopping_list.py:
```diff
@@ -175,23 +215,34 @@ def async_update(self, item_id, info):
-    def save(self):
+    async def async_load(self):
+        """Load lists."""
+        self.lists = await self.hass.helpers.storage.async_migrator(
```
Sorry,  I don't follow on this one

iantrich replied:
@balloob are you able to clarify this at all for me? 🙏

balloob replied:
To avoid confusion, let me use the word `array` instead of Python's data structure `list`.

In the storage component, you store an array of lists. You keep the same data structure when loaded into memory. This means that for you to look up a shopping list, you're doing a `next(…)` call to extract the list from the array.

Instead, in memory, you want to convert the array of lists to a dictionary, keyed by the ID of the list. That way you're able to do fast look ups. You will however need to preserve the order the lists are created, as that might be significant. Standard dictionaries are only ordered starting Python 3.6, so you need to use `OrderedDict`.

```python
from collections import OrderedDict

data = await self.hass.helpers.storage.async_migrator(…)
lists = OrderedDict()
if data is None:
    …

for list in data:
    lists[list['id']] = list

self.lists = lists
```

Then update `_data_to_save` to convert the dictionary to a list again: `list(self.lists.values())`

iantrich committed:
Fix test

tox is complaining about `slugify` module not existing. I might have broken something...not sure

iantrich commented:
closing as a new task domain is explored
